src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb

```diff
@@ -1,250 +1,24 @@
-            result = (model, model.SyntaxTree.GetRoot())
-            Dim invertedIfNode = result.Root.GetAnnotatedNodesAndTokens(s_ifNodeAnnotation).Single().AsNode()
-
-            Dim explicitInvertedIfNode = Simplifier.Expand(invertedIfNode, result.Model, document.Project.Solution.Workspace, cancellationToken:=cancellationToken)
-            result = UpdateSemanticModel(result.Model, result.Root.ReplaceNode(invertedIfNode, explicitInvertedIfNode), cancellationToken)
-
-            Return result.Root
-        End Function
-
-        Private Function UpdateSemanticModel(model As SemanticModel, root As SyntaxNode, cancellationToken As CancellationToken) As (Model As SemanticModel, Root As SyntaxNode)
-            Dim newModel = model.Compilation.ReplaceSyntaxTree(model.SyntaxTree, root.SyntaxTree).GetSemanticModel(root.SyntaxTree)
-            Return (newModel, newModel.SyntaxTree.GetRoot(cancellationToken))
-        End Function
-
-        Private Function InvertSingleLineIfStatement(
-                document As Document,
-                originalIfNode As SingleLineIfStatementSyntax,
-                generator As SyntaxGenerator,
-                syntaxFacts As ISyntaxFactsService,
-                model As SemanticModel,
-                cancellationToken As CancellationToken) As SemanticModel
-            Dim root = model.SyntaxTree.GetRoot()
-            Dim invertedIfNode = GetInvertedIfNode(originalIfNode, document, generator, syntaxFacts, model, cancellationToken)
-            Dim result = UpdateSemanticModel(model, root.ReplaceNode(originalIfNode, invertedIfNode), cancellationToken)
-
-            ' Complexify the next statement if there is one.
-            invertedIfNode = DirectCast(result.Root.GetAnnotatedNodesAndTokens(s_ifNodeAnnotation).Single().AsNode(), SingleLineIfStatementSyntax)
-
-            Dim currentStatement As StatementSyntax = invertedIfNode
-            If currentStatement.HasAncestor(Of ExpressionSyntax)() Then
-                currentStatement = currentStatement _
-                    .Ancestors() _
-                    .OfType(Of ExpressionSyntax) _
-                    .Last() _
-                    .FirstAncestorOrSelf(Of StatementSyntax)()
-            End If
-
-            Dim nextStatement = currentStatement.GetNextStatement()
-            If nextStatement IsNot Nothing Then
-                Dim explicitNextStatement = Simplifier.Expand(nextStatement, result.Model, document.Project.Solution.Workspace, cancellationToken:=cancellationToken)
-                result = UpdateSemanticModel(result.Model, result.Root.ReplaceNode(nextStatement, explicitNextStatement), cancellationToken)
-            End If
-
-            Return result.Model
+        Protected Overrides Function GetIfStatement(token As SyntaxToken) As SyntaxNode
+            Return If(DirectCast(token.GetAncestor(Of SingleLineIfStatementSyntax), SyntaxNode),
+                      DirectCast(token.GetAncestor(Of MultiLineIfBlockSyntax), SyntaxNode))
```

Comment:
would it potentially be cleaner to have two separate analyzers for this sort of thing.  That way you could be very strongly typed over single-line vs multi-line if blocks, and you woudln't have to use SyntaxNode as the common type that VB operates over.

Reply:
(resolved)